internal/bson2/logging.go

```diff
@@ -0,0 +1,273 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bson2
+
+import (
+	"encoding/base64"
+	"encoding/hex"
+	"fmt"
+	"log/slog"
+	"math"
+	"strconv"
+	"strings"
+	"time"
+)
+
+// flowLimit is the maximum length of a flow/inline/compact representation of a BSON value.
+// It may be set to 0 to disable flow representation.
+const flowLimit = 80
+
+// nanBits is the most common pattern of a NaN float64 value, the same as math.Float64bits(math.NaN()).
+const nanBits = 0b111111111111000000000000000000000000000000000000000000000000001
+
+// slogValue returns a compact representation of any BSON value as [slog.Value].
+// It may change over time.
+//
+// The result is optimized for small values such as function parameters.
+// Some information is lost;
+// for example, both int32 and int64 values are returned with [slog.KindInt64],
+// arrays are treated as documents, and empty documents are omitted.
+// More information is subsequently lost in handlers output;
+// for example, float64(42), int32(42), and int64(42) values would all look the same
+// (`f64=42 i32=42 i64=42` or `{"f64":42,"i32":42,"i64":42}`).
```

Comment:
Given all those limitations, I'm not sure if we really need that. Let's keep it for now, but it is very possible it will be removed soon.